src/inference/tempering.jl

```diff
@@ -0,0 +1,178 @@
+###########################
+#   SAMPLING
+###########################
+
+"""
+Upon calling sample on a `TemperedSampler`, we must manually instantiate a `Sampler`
+to insert into the `TemperedSampler`, rather than wrapping the `TemperedSampler`
+itself in a `Sampler`
+"""
+function AbstractMCMC.sample(
+    rng::AbstractRNG,
+    model::AbstractModel,
+    sampler::MCMCTempering.TemperedSampler{<:InferenceAlgorithm},
+    N::Integer;
+    chain_type=MCMCChains.Chains,
+    resume_from=nothing,
+    progress=PROGRESS[],
+    kwargs...
+)
+    sampler = MCMCTempering.TemperedSampler(
+        Sampler(sampler.internal_sampler, model),
+        sampler.Δ,
+        sampler.Δ_init,
+        sampler.N_swap,
+        sampler.swap_strategy
+    )
+
+    if resume_from === nothing
+        return AbstractMCMC.mcmcsample(rng, model, sampler, N;
+                                       chain_type=chain_type, progress=progress, kwargs...)
+    else
+        return resume(resume_from, N; chain_type=chain_type, progress=progress, kwargs...)
+    end
+end
+
+
+function AbstractMCMC.sample(
+    rng::AbstractRNG,
+    model::AbstractModel,
+    sampler::MCMCTempering.TemperedSampler{<:InferenceAlgorithm},
+    ensemble::AbstractMCMC.AbstractMCMCEnsemble,
+    N::Integer,
+    n_chains::Integer;
+    chain_type=MCMCChains.Chains,
+    progress=PROGRESS[],
+    kwargs...
+)
+    sampler = MCMCTempering.TemperedSampler(
+        Sampler(sampler.internal_sampler, model),
+        sampler.Δ,
+        sampler.Δ_init,
+        sampler.N_swap,
+        sampler.swap_strategy
+    )
+
+    return AbstractMCMC.mcmcsample(rng, model, sampler, ensemble, N, n_chains;
+                                   chain_type=chain_type, progress=progress, kwargs...)
+end
+
+
+function AbstractMCMC.bundle_samples(
+    ts::Vector,
+    model::DynamicPPL.Model,
+    spl::MCMCTempering.TemperedSampler,
+    state::MCMCTempering.TemperedState,
+    chain_type::Union{Type{MCMCChains.Chains},Type{Vector{NamedTuple}}};
+    kwargs...
+)
+    return AbstractMCMC.bundle_samples(ts, model, spl.internal_sampler, state, chain_type; kwargs...)
+end
+
+
+
+###########################
+#   MODEL
+###########################
+
+struct TemperedEval{T<:AbstractFloat}
+    model :: DynamicPPL.Model
+    β     :: T
+end
+
+function (te::TemperedEval)(
+    rng::Random.AbstractRNG,
+    model::DynamicPPL.Model,
+    varinfo::DynamicPPL.AbstractVarInfo,
+    sampler::DynamicPPL.AbstractSampler,
+    context::DynamicPPL.AbstractContext,
+    args...
+)
+    context = DynamicPPL.MiniBatchContext(
+        context,
+        te.β
+    )
+    te.model.f(rng, model, varinfo, sampler, context, args...)
```

Comment:
I think this is dangerous - `model.f` is a very internal thing and we even avoid calling it explicitly in DynamicPPL in most places. We explicitly wrap it in https://github.com/TuringLang/DynamicPPL.jl/blob/5ffbbd235a1638b4b231bd1741e55c75e83215cc/src/model.jl#L150-L152 to ensure that the types of the inputs are correct, and even this function is not called directly but rather we use `model(...)` which ensures that evaluation is performed in a threadsafe way: https://github.com/TuringLang/DynamicPPL.jl/blob/5ffbbd235a1638b4b231bd1741e55c75e83215cc/src/model.jl#L85-L89

Reply:
In any case, it seems surprising to use `te.model.f` instead of `model.f`?

Reply:
Yes, this was our main concern with the approach to inject a context for Turing samplers, is there an obvious alternative here to achieve the same thing that would be safer? We concluded that as it was at least some external functionality it might be ok to have it work as it does.

As for your second question, I believe the call to te.model.f is required to avoid an infinite loop, as in this case the `Model`'s `F` is just the TemperedEval  call itself. We define a new model object with this wrapper but store the original model within it to be able to just alter the context and then call the original `model.f` with a new context via `te.model.f`, as to call  just `model.f` would be to precisely call on the `TemperedEval` object, hopefully that makes sense?

Reply:
i.e. the `model` in the args of `function (te::TemperedEval)` is the output of `make_tempered_model`, not the original model with functional `f` that we want to evaluate

Reply:
This is the last outstanding question for this PR, @devmotion @mohamed82008 any idea on how best to move forward with this? Also, thanks for all the help + comments David :)

Reply:
Ah I missed your replies here, let me check the implementation again.

BTW there's still https://github.com/TuringLang/Turing.jl/pull/1628#discussion_r642933638 that has to be resolved :slightly_smiling_face: